src/Controller/MakefileToJobTree.ts

```diff
@@ -0,0 +1,63 @@
+import { basename, dirname } from 'path'
+import { exec } from 'child_process'
+import { promisify } from 'util'
+import { Job } from '..'
+import { makefileTraceToJobTree } from './MakefileTraceToJobTree'
+import { MakeTracingError } from './MakeTracingError'
+
+/**
+ * Gets a trace from a makefile by running make with the --trace and --dry-run options.
+ *
+ * @param filePath - the path to the Makefile
+ * @returns the trace.
+ */
+async function getMakefileTrace(options: {
+	filePath?: string
+	targets?: string[]
+}): Promise<string> {
+	let command = 'make --trace --dry-run'
+
+	if (options.filePath) {
+		const fileName = basename(options.filePath)
+		const directoryName = dirname(options.filePath)
+
+		command += ` -f "${fileName}" --directory="${directoryName}"`
+	}
+
+	if (options.targets) {
+		command += ` ${options.targets.join(' ')}`
+	}
+
+	let stdout: string
+	let stderr: string
+	try {
+		;({ stdout, stderr } = await promisify(exec)(command))
+	} catch (error) {
+		throw new MakeTracingError(
+			`Error while generating make trace using command "${command}":\n${error}`,
+		)
+	}
+
+	if (stderr.length > 0) {
+		throw new MakeTracingError(
+			`Error while generating make trace using command "${command}".\n\nStderr:\n${stderr}\n\nStdout:${stdout}`,
```

Comment:
Couldn't `stderr` and `stdout` end up being massive strings? That would be a pretty unhelpful error message. It's better to add those as fields in the error object (and extend the ctor to accept those fields).

Reply:
What should I do with the fields? Just leave them there in case someone wants to use them in the future?

Reply:
I believe they will get printed out whenever the error is console.logged or similar. You don't need to do anything with the fields aside from storing them.